### PR TITLE
*: add SplitRegion API

### PR DESCRIPTION
**Pull request by overvenus**
CC #2270, pingcap/kvproto#191

### CODE REVIEW

File: src/server/grpc_service.rs [hunk 1]
```diff
@@ -931,6 +932,54 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
...
+        sink: UnarySink<SplitRegionResponse>,
+    ) {
+        let label = "split_region";
+        let timer = GRPC_MSG_HISTOGRAM_VEC
+            .with_label_values(&[label])
+            .start_coarse_timer();
+
+        let (cb, future) = make_callback();
+        let req = StoreMessage::SplitRegion {
+            split_key: req.take_split_key(),
```
Comment (disksing) [inline src/server/grpc_service.rs]: Need to encode the split key.

Comment (overvenus) [reply src/server/grpc_service.rs]: Sure!

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1813,32 +1828,88 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        split_key: Vec<u8>,
+        region_id: Option<u64>,
+        epoch: Option<metapb::RegionEpoch>,
+        cb: Option<Callback>,
+    ) {
+        let region_id = match region_id {
+            Some(id) => id,
+            None => match self.region_ranges
+                .range((Excluded(split_key.clone()), Unbounded::<Key>))
+                .next()
```
Comment (disksing) [inline src/raftstore/store/store.rs]: Need to check if the region contains `split_key`?
btw, is it ok if `split_key == region.start_key`?

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -1813,32 +1828,88 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+        &mut self,
+        split_key: Vec<u8>,
+        region_id: Option<u64>,
+        epoch: Option<metapb::RegionEpoch>,
+        cb: Option<Callback>,
+    ) {
+        let region_id = match region_id {
+            Some(id) => id,
+            None => match self.region_ranges
+                .range((Excluded(split_key.clone()), Unbounded::<Key>))
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: How about split_key == someone region's end_key?

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (siddontang) [reply src/raftstore/store/store.rs]: seem we still need to check the split key validation like the split coprocessor does.

File: src/raftstore/store/msg.rs [hunk 4]
```diff
@@ -67,9 +67,16 @@ pub enum Msg {
     SplitCheckResult {
         region_id: u64,
         epoch: RegionEpoch,
+        // It's a data key, starts with `DATA_PREFIX_KEY`.
```
Comment (siddontang) [inline src/raftstore/store/msg.rs]: We may support meta key later.

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -1838,49 +1853,148 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            Some(region_id),
+            Some(epoch),
+            None,
+        );
+    }
+
+    fn on_prepare_split_region(
+        &mut self,
+        split_key: Vec<u8>, // `split_key` is a encoded key.
+        region_id: Option<u64>,
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: We must know the region id here. Don't allow the user send a request without a region id.

Comment (siddontang) [reply src/raftstore/store/store.rs]: E.g, if the user creates a table and wants to split the region, he/she must find the correct the region, then sends the split request.

File: src/raftstore/store/store.rs [hunk 6]
```diff
@@ -1838,49 +1853,148 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                        split_key,
+                        e
+                    );
+                    cb.map(|cb| {
+                        cb(new_error(
+                            box_err!("failed to split at {:?}: {}", split_key, e),
+                        ))
+                    });
+                } else {
+                    entry.insert(cb);
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: we should pass the callback to the PD and let PD worker passes it as the split request callback to the raftstore.

Assume we have following cases:

1. AskSplit and insert to callback hash map.
2. PD workers sends a split request to raft store
3. The peer became follower, so the split request can be executed
4. The callback in the hash map can be called.

File: src/raftstore/store/store.rs [hunk 7]
```diff
@@ -1828,55 +1828,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+                    region_id,
+                    self.store_id()
+                );
+                cb.map(|cb| cb(new_error(Error::RegionNotFound(region_id))));
+                return;
+            }
+            Some(peer) => {
+                if !peer.is_leader() {
+                    // region on this store is no longer leader, skipped.
+                    error!(
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: info is enough.

File: src/raftstore/store/store.rs [hunk 8]
```diff
@@ -1828,55 +1828,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            }
+        };
 
-        let peer = p.unwrap();
         let region = peer.region();
 
         if region.get_region_epoch().get_version() != epoch.get_version() {
-            info!(
-                "{} epoch changed {:?} != {:?}, need re-check later",
+            error!(
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: info

File: tests/raftstore/test_split_region.rs [hunk 9]
```diff
@@ -101,28 +106,56 @@ fn test_base_split_region<T: Simulator>(cluster: &mut Cluster<T>, right_derive:
...
+
+#[test]
+fn test_server_manual_split_region_right_derive() {
+    let count = 5;
+    let mut cluster = new_server_cluster(0, count);
+    test_base_split_region(&mut cluster, Cluster::must_manual_split, false);
+}
+
+#[test]
+fn test_server_manual_split_region_left_derive() {
```
Comment (siddontang) [inline tests/raftstore/test_split_region.rs]: do we have a test to send multi Split requests with the same split key? Maybe only the first can be executed successfully.

Comment (siddontang) [reply tests/raftstore/test_split_region.rs]: any update?

**Comment by overvenus** [comment]
PTAL @siddontang @BusyJay @disksing

File: tests/raftstore/test_split_region.rs [hunk 10]
```diff
@@ -101,28 +107,92 @@ fn test_base_split_region<T: Simulator>(cluster: &mut Cluster<T>, right_derive:
...
+        let right = split_resp.take_right();
+        assert_eq!(left.get_end_key(), key.as_slice());
+        assert_eq!(region2.get_start_key(), left.get_start_key());
+        assert_eq!(left.get_end_key(), right.get_start_key());
+        assert_eq!(region2.get_end_key(), right.get_end_key());
+    });
+    cluster.split_region_by_key(&region, split_key, Some(c));
+    let c = Box::new(move |resp: RaftCmdResponse| {
+        assert!(resp.has_header());
+        assert!(resp.get_header().has_error());
```
Comment (siddontang) [inline tests/raftstore/test_split_region.rs]: what is the error? stale epoch?

Comment (overvenus) [reply tests/raftstore/test_split_region.rs]: Now, it is `invalid split request: ...`, originates at apply.rs#L862.

https://github.com/pingcap/tikv/blob/61eb812de8df59a23a210df2776125af8ad5b6b0/src/raftstore/store/worker/apply.rs#L862

**Comment by overvenus** [comment]
PTAL

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay @javaforfun

File: src/raftstore/store/store.rs [hunk 11]
```diff
@@ -1828,55 +1828,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-            info!(
-                "[region {}] region on {} doesn't exist or is not leader, skip.",
-                region_id,
-                self.store_id()
-            );
-            return;
-        }
+        let peer = match self.region_peers.get(&region_id) {
+            None => {
+                error!(
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Can be info too.

File: src/raftstore/store/store.rs [hunk 12]
```diff
@@ -1828,55 +1828,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            }
+        };
 
-        let peer = p.unwrap();
         let region = peer.region();
 
         if region.get_region_epoch().get_version() != epoch.get_version() {
             info!(
-                "{} epoch changed {:?} != {:?}, need re-check later",
+                "[region {}] {} epoch changed {:?} != {:?}, need re-check later",
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: peer.tag should contain `[region {}]` already.

File: src/raftstore/store/store.rs [hunk 13]
```diff
@@ -1828,55 +1828,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
     }
 
-    fn on_split_check_result(
+    fn on_prepare_split_region(
         &mut self,
         region_id: u64,
         epoch: metapb::RegionEpoch,
-        split_key: Vec<u8>,
+        split_key: Vec<u8>, // `split_key` is a encoded key.
+        cb: Option<Callback>,
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: Can `cb` be handled outside the method? So there won't be so many places that need to expand cb.

File: src/raftstore/store/store.rs [hunk 14]
```diff
@@ -1828,55 +1828,99 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
         }
 
-        let key = keys::origin_key(&split_key);
         let task = PdTask::AskSplit {
             region: region.clone(),
-            split_key: key.to_vec(),
+            split_key: split_key.clone(),
             peer: peer.peer.clone(),
             right_derive: self.cfg.right_derive_when_split,
+            callback: if let Some(cb) = cb {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: You mean `unwrap_or_else`?

File: tests/raftstore/test_split_region.rs [hunk 15]
```diff
@@ -101,28 +108,97 @@ fn test_base_split_region<T: Simulator>(cluster: &mut Cluster<T>, right_derive:
...
 #[test]
 fn test_server_base_split_region_right_derive() {
     let count = 5;
     let mut cluster = new_server_cluster(0, count);
-    test_base_split_region(&mut cluster, true);
+    test_base_split_region(&mut cluster, Cluster::must_split, true);
+}
+
+#[test]
+fn test_node_manual_split_region_right_derive() {
```
Comment (BusyJay) [inline tests/raftstore/test_split_region.rs]: Too many test cases. Actually just one should be ok.

File: src/raftstore/store/msg.rs (hunk 4 above)
Comment (javaforfun) [reply src/raftstore/store/msg.rs]: can be same type with SplitRegion.split_key

File: src/raftstore/store/worker/pd.rs [hunk 16]
```diff
@@ -493,7 +489,14 @@ fn send_admin_request_raw(
 
     req.set_admin_request(request);
 
-    if let Err(e) = ch.try_send(Msg::new_raft_cmd(req, Box::new(|_| {}))) {
+    if let Err(e) = ch.try_send(Msg::new_raft_cmd(
+        req,
+        if let Some(cb) = callback {
```
Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: Use `unwrap_or_else` instead.

File: src/server/service/kv.rs [hunk 17]
```diff
@@ -925,6 +926,54 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
...
+        if let Err(e) = self.ch.try_send(req) {
+            self.send_fail_status(ctx, sink, Error::from(e), RpcStatusCode::ResourceExhausted);
+            return;
+        }
+
+        let future = future
+            .map_err(Error::from)
+            .map(|mut v| {
+                let mut resp = SplitRegionResponse::new();
+                if v.has_header() && v.get_header().has_error() {
```
Comment (BusyJay) [inline src/server/service/kv.rs]: `has_header` is unnecessary.

File: src/raftstore/store/store.rs [hunk 18]
```diff
@@ -2389,13 +2423,19 @@ impl<T: Transport, C: PdClient> mio::Handler for Store<T, C> {
...
                 region_id,
                 epoch,
                 split_key,
             } => {
                 info!("[region {}] split check complete.", region_id);
-                self.on_split_check_result(region_id, epoch, split_key);
+                self.on_prepare_split_region(
+                    region_id,
+                    epoch,
+                    keys::origin_key(split_key.as_slice()).to_vec(),
```
Comment (javaforfun) [inline src/raftstore/store/store.rs]: pd AskSplit use different split_key type now?

**Comment by javaforfun** [comment]
rest lgtm

File: src/raftstore/store/store.rs (hunk 18 above)
Comment (overvenus) [reply src/raftstore/store/store.rs]: Actully the `split_key` here is a data key, AskSplit requires encoded key, so convert it by `keys::origin_key`.

File: tests/raftstore/cluster.rs [hunk 19]
```diff
@@ -734,7 +734,35 @@ impl<T: Simulator> Cluster<T> {
...
 
-    pub fn ask_split(&mut self, region: &metapb::Region, split_key: &[u8]) {
+    // It's similar to `ask_split`, the difference is the msg, it sends, is `Msg::SplitRegion`,
+    // and `region` will not be embedded to that msg.
+    // Caller must ensure that the `split_key` is in the `region`.
+    pub fn split_region_by_key(
+        &mut self,
+        region: &metapb::Region,
+        split_key: &[u8],
+        cb: Option<Callback>,
```
Comment (BusyJay) [inline tests/raftstore/cluster.rs]: Why `Option`?

File: tests/raftstore/test_split_region.rs [hunk 20]
```diff
@@ -101,28 +108,76 @@ fn test_base_split_region<T: Simulator>(cluster: &mut Cluster<T>, right_derive:
...
+        let right = split_resp.take_right();
+        assert_eq!(left.get_end_key(), key.as_slice());
+        assert_eq!(region2.get_start_key(), left.get_start_key());
+        assert_eq!(left.get_end_key(), right.get_start_key());
+        assert_eq!(region2.get_end_key(), right.get_end_key());
+        tx.send(right).unwrap();
+    });
+    cluster.split_region_by_key(&region, split_key, Some(c));
+
+    let region3 = rx.recv().unwrap();
```
Comment (BusyJay) [inline tests/raftstore/test_split_region.rs]: Should `recv_timeout`.

File: tests/raftstore/test_split_region.rs [hunk 21]
```diff
@@ -101,28 +108,76 @@ fn test_base_split_region<T: Simulator>(cluster: &mut Cluster<T>, right_derive:
...
+    cluster.must_put(split_key, b"v2");
+    let c = Box::new(move |resp: RaftCmdResponse| {
+        assert!(resp.has_header());
+        assert!(resp.get_header().has_error());
+        assert!(!resp.has_admin_response());
+    });
+    cluster.split_region_by_key(&region3, split_key, Some(c));
```
Comment (BusyJay) [inline tests/raftstore/test_split_region.rs]: The test may exit before executing callback.

**Comment by siddontang** [comment]
Any update?

**Comment by overvenus** [comment]
PTAL

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**